Add support for userspace services (#4869)
tock/tock: capsules/system/src/userspace_services/README.md

# Overview

> The userspace service architecture enables application-layer code to offer service-level functionality to clients with a structured interface for invoking operations and communicating data. This allows both the service and the client to evolve and change independent of each other due to abiding by a well-defined interface.

Reviewer: ```suggestion
The userspace service architecture enables application-layer code to offer kernel HIL implementations to clients with a structured interface for invoking operations and communicating data. This allows both the service and the client to evolve and change independent of each other due to abiding by a well-defined interface.
```

"service" is underdefined here, and I think it implies various things to various people depending on that context they're assuming.

> Applications offering service-level functionality, userspace service applications, are available for use by the entire platform and run as application-layer code that is individually deployable just as other applications are.

Reviewer: ```suggestion
Applications offering kernel HIL implementations, userspace service applications, are available for use by the entire platform and run as application-layer code that is individually deployable just as other applications are.
```

> Applications offering service-level functionality, userspace service applications, are available for use by the entire platform and run as application-layer code that is individually deployable just as other applications are.

Reviewer: What does it mean to be "available for use by the entire platform"? Actually only capsules can call into these right? Or other kernel mechanisms, I guess. Those capsules could themselves be triggered by other applications or by whatever. But really these are only directly callable from within the kernel.

> The architecture achieves a decoupling of service code from both the OS and the consumer of the service.

Reviewer: ```suggestion
The architecture achieves a decoupling of service code from both the OS and the consumer of the service by relying on the stability of the HIL trait.
```

Changes in the userspace service application do not cascade into requiring changes to client code.
Changes in the userspace service application code do not require an OS-level modification or update.

## Use case examples

- Swappable suites of cryptographic primitives, promoting cryptographic agility.
- Sensor fusion and soft sensing, synthesizing data from multiple sources as a new stream of data.
- Network stacks, connecting multiple applications while keeping networking logic out of the kernel.
- Data storage, providing a structured, on-device data store.

# Architecture

There are three major components to the userspace services architecture:

> - the **registry**, which tracks and mediates communication with all userspace service applications;

Reviewer: ```suggestion
- the **registry**, a capsule which tracks and mediates communication with all userspace service applications;
```

- the **userspace service application**, which offers service-level functionality from userspace;

> - and the **service interface**, which maps HIL functions to calls to the userspace application.

Reviewer: ```suggestion
- and the **service interface**, a capsule which maps HIL functions to calls to the userspace application.
```

                                                 |        | |
              +---------+                        |        | |
              |HIL trait|                        |        | |
         +----+---------+------------------+     |        | |
         |        Service Interface        |     |        | |
         +---------------------------------+     |        | |
              |UserspaceServiceClient trait|     |        | |
              +----------------------------+     |        | |
                                                 |        | |
                                                 |        | |
         +----------------------------+          |        | |
         |UserspaceServiceAccess trait|          v        | v
         +----------------------------------------------------------+
         |                          Registry                        |
         +----------------------------------------------------------+
```

At startup, the userspace service application registers with the Registry with the register command syscall which also identifies the userspace service application’s role with a role ID.
This registration is only necessary once, at startup.

> Its return value indicates the success or failure of the registration.

Reviewer: What happens if a request is made before the userspace service registers?

    =================================================================
      KERNEL

         +--------------------+
         |      Capsule       |
         +--------------------+

              +---------+
              |HIL trait|
         +----+---------+------------------+
         |        Service Interface        |
         +---------------------------------+
            | |UserspaceServiceClient trait|
            | +----------------------------+
 usercall() |                            ^
            v            usercall_done() |
         +----------------------------+  |
         |UserspaceServiceAccess trait|  |
         +----------------------------------------------------------+
         |                          Registry                        |
         +----------------------------------------------------------+
```

Upon receiving a function call from its HIL client, the service interface maps arguments received from the HIL client to arguments to pass to the userspace service (through the Registry).
The service interface achieves this by either serializing bytes (for intrinsic data types that fit in a `usize`)

> or passing the buffers containing data received from the HIL client.

Reviewer: Those buffers are going to have to be copied to/from the userspace service too, right? I guess that should be explained in the above section.

The Registry receives this data through a call to the `usercall` function on the `UserspaceServiceAccess` trait.

> When a userspace service operation completes, the Registry makes a callback to the service interface through the `UserspaceServiceClient` trait function `usercall_done`.

Reviewer: What does the userspace service do that connects to the usercall_done? That's separate from the command that signals success/failure, right? (again, I think this goes in the prior subsection, but I didn't think of the question until I saw the text here)
